File: Java/boatparts.java

// https://open.kattis.com/problems/boatparts
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.TreeSet;

public class boatparts {
    public static void main(String[] args) throws IOException {
        BufferedReader sc = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter dc = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] line = sc.readLine().split(" ");
        int totalparts = Integer.parseInt(line[0]);
        int days = Integer.parseInt(line[1]);
        boolean found = false;
        TreeSet<String> parts = new TreeSet<>();
        for (int x = 1; x <= days; x++) {
            if (parts.size() != totalparts)
                parts.add(sc.readLine());
            if (parts.size() == totalparts) {
                dc.write(x + "\n");
                found = true;
                break;
            }
        }
        if (!found)
            dc.write("paradox avoided\n");
        dc.flush();
        sc.close();
        dc.close();
    }
}

File: C Sharp/boatparts.cs
using System;
using System.Collections.Generic;

public class boatparts {
    public static void Main (string[] args) {
        string[] line = Console.ReadLine ().Split (' ');
        int total = int.Parse (line[0]);
        int days = int.Parse (line[1]);
        bool found = false;
        SortedSet<string> parts = new SortedSet<string> ();
        for (int x = 1; x <= days; x++) {
            parts.Add (Console.ReadLine ());
            if (parts.Count == total) {
                Console.WriteLine (x);
                found = true;
                break;
            }
        }
        if (!found)
            Console.WriteLine ("paradox avoided");
    }
}